### PR TITLE
[extractor/youtube] Add compat-opt to revert "Prefer UTC upload date for videos"

### PR DESCRIPTION
**Pull request by coletdjnz**
**IMPORTANT**: PRs without the template will be CLOSED

### Description of your *pull request* and other information

</details>

ADD DESCRIPTION HERE

Adds `--compat-options no-youtube-prefer-utc-upload-date` (open to suggestions for better name).

This reverts https://github.com/yt-dlp/yt-dlp/commit/992f9a730b49fd36fc422be8d802f98ebcdce418 and https://github.com/yt-dlp/yt-dlp/commit/17322130a954577bb03b833d5c435638e51e19f2 to prefer the non-UTC upload date in microformats. 

This is needed in some cases where the preference of the UTC date is problematic (e.g. compatibility reasons or us switching between UTC and non-UTC depending on if the video is a live stream etc.)

<details open><summary>Template</summary> 

### Before submitting a *pull request* make sure you have:
- [X] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [X] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [X] Checked the code with [flake8](https://pypi.python.org/pypi/flake8) and [ran relevant tests](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [X] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [X] Fix or improvement to an extractor (Make sure to add/update tests)
- [ ] New extractor ([Piracy websites will not be accepted](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#is-the-website-primarily-used-for-piracy))
- [ ] Core bug fix/improvement
- [ ] New feature (It is strongly [recommended to open an issue first](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#adding-new-feature-or-making-overarching-changes))

### CODE REVIEW

**Comment by pukkandan** [comment]
> This is needed in some cases where the preference of the UTC date is problematic (e.g. compatibility reasons or us switching between UTC and non-UTC depending on if the video is a live stream etc.)

The timestamp was always meant to be UTC. It being not so in previous versions is a bug. So compatibility is not an issue here. Can you explain your usecase?

**Comment by ivan** [comment]
I suggested this feature earlier because I have a large number of YouTube captures (going back to like 2010) with the non-UTC `upload_date` and I sometimes capture the same video again; having different `upload_date`s for the same video ID is very undesired for me. I carefully maintained things to avoid the UTC date and I would have to carry my own patch forever without this feature.

File: README.md [hunk 1]
```diff
@@ -141,6 +141,7 @@ Some of yt-dlp's default options are different from that of youtube-dl and youtu
...
 * Youtube channel URLs are automatically redirected to `/video`. Append a `/featured` to the URL to download only the videos in the home page. If the channel does not have a videos tab, we try to download the equivalent `UU` playlist instead. For all other tabs, if the channel does not show the requested tab, an error will be raised. Also, `/live` URLs raise an error if there are no live videos instead of silently downloading the entire channel. You may use `--compat-options no-youtube-channel-redirect` to revert all these redirections
 * Unavailable videos are also listed for youtube playlists. Use `--compat-options no-youtube-unavailable-videos` to remove this
+* The upload date in UTC timezone is preferred over the non-UTC upload date for YouTube videos in [most circumstances](https://github.com/yt-dlp/yt-dlp/blob/master/yt_dlp/extractor/youtube.py#L3898-L3900). Use `--compat-options no-youtube-prefer-utc-upload-date` to prefer the non-UTC upload date.
```
Comment (pukkandan) [inline README.md:144]:
```suggestion
* The upload dates extracted from YouTube are in UTC [when available](https://github.com/yt-dlp/yt-dlp/blob/89e4d86171c7b7c997c77d4714542e0383bf0db0/yt_dlp/extractor/youtube.py#L3898-L3900). Use `--compat-options no-youtube-prefer-utc-upload-date` to prefer the non-UTC upload date.
```

**Review by pukkandan — COMMENTED** [review]
Name suggestions:
`no-youtube-utc-upload-date`
`no-youtube-utc`

**Comment by coletdjnz** [comment]
> Name suggestions: `no-youtube-utc-upload-date` `no-youtube-utc`

`no-youtube-utc` is too broad imo.

`no-youtube-utc-upload-date` could work

**Comment by pukkandan** [comment]
Any reason this is pending?

**Comment by coletdjnz** [comment]
> Any reason this is pending?

was unsure about the opt name. but I think how it is rn is fine

**Comment by dirkf** [comment]
Surely the problem here is actually `unified_strdate()`? It deliberately removes and discards AM/PM and TZ information from the date string.

Once upon a time YT may have tagged videos with just YYYYMMDD in PST, so this would have been acceptable. Now the data sent is ISO 8601 TZ format, both in JSON data and in page HTML.

Therefore the extractor should be using `unified_timestamp()`, which extracts **and uses** any AM/PM and XXX TZ data, or just `parse_iso8601()`, pending some decision on how or whether to fix `unified_strdate()`. I believe it should just return the value of `unified_timestamp()` formatted as YYYMMDD relative to UTC; a compat option could format the string relative to any original TZ.

Eg, YouTube test video 4L2J27mJ3Dc, with yt-dl extractor:
```console
(Pdb) p microformat.get('uploadDate')
'2022-01-02T16:48:59-08:00'
(Pdb) p search_meta('uploadDate')
'2022-01-02T16:48:59-08:00'
(Pdb) p unified_timestamp(microformat.get('uploadDate'))
1641170939
(Pdb) 
```
And
```console
$ date --date='@1641170939' -u
Mon  3 Jan 00:48:59 UTC 2022
$ 
```
Old-style result: 20220102
New-style result: 20220103 (by ~49 minutes)

**Comment by coletdjnz** [comment]
When this (and the prior PRs) were written, youtube was only returning the date as YYYYMMDD in PST. If that has since changed and the metadata is being extracted wrong then a new issue should be raised.

cheers

**Pull request merged**